[RF-DOCS] Update the JavaScript in Rails guide (#58165)
rails/rails: guides/source/working_with_javascript_in_rails.md

Working with JavaScript in Rails
================================

This guide covers the integration of JavaScript into your Rails application —
including the usage of Turbo and Stimulus as well as the installation of
external JavaScript packages into Rails.

After reading this guide, you will know:

- How to use an Import Map to deliver JavaScript in your Rails app.
- How to integrate JavaScript bundlers like `esbuild` or `rollup` with Rails.
- What Turbo is and how Rails integrates with it.
- How to install Stimulus and use it for client-side JavaScript functionality.
- How to make HTTP requests using JavaScript with the `request.js` library.

--------------------------------------------------------------------------------

Introduction
------------

> Rails provides two mechanisms to deliver JavaScript within your application:

Reviewer: You can still require jquery in a vanilla Rails application to deliver JavaScript:
```
<script src="app/assets/jquery.min.js"></script>
```

So I think this sentence is either too strict or it should mention ES modules (or npm packages).
Maybe something like (some copy borrowed from the importmaps-rails readme):
```suggestion
ES modules (ESM) is the standard format for packaging JavaScript code in modern applications. Rails provides two mechanisms to deliver JavaScript 
made for ESM:
```

using an [import map](#using-a-javascript-import-map), or a
[JavaScript bundler](#using-a-javascript-bundler). Both of these systems
integrate with the [Asset Pipeline](asset_pipeline.html) to deliver the files to
the browser.

The [Turbo](#turbo) and [Stimulus](#stimulus) JavaScript libraries (which are
part of the [Hotwire](https://hotwired.dev) suite) are included in Rails, and

> form the default front-end stack.

Reviewer: This guide used to describe how to use JavaScript to add interactivity to an application. With the new Hotwire guide this guide gets another focus: a more general guide to managing JavaScript in a Rails application.

I still think this guide will be the starting point for readers wanting to add interactivity to their application. The Rails convention is using Turbo and maybe adding some Stimulus controllers for extra functionality. A reader generating a new Rails application could probably care less how the JavaScript is required, and care more how to make things more interactive.

What do you think about starting this section with Turbo and Stimulus (basically moving this paragraph to the top)? 
Starting with Hotwire maybe somewhat counterintuitive when we have a Hotwire guide. But I think we can be opinionated here, and start from the assumption that the reader should probably use Hotwire for a new application, instead of using importmap-rails to install React or Vue.

Author: I'd like to keep this guide a little more generic so I'd prefer to keep this paragraph where it is. 

I see this guide more as a reference to do something advanced with JavaScript. With the Hotwire guide in the pipeline, I'd like to start this guide by talking about JavaScript in general, rather than specific libraries.

Reviewer: Yeah, this guide feels somewhat like a grab bag of topics: requiring JavaScript ES modules, using Hotwire, and request.js.
With the Hotwire guide this leaves basically requiring JavaScript ES modules and request.js, which are two specific loosely related topics.

Starting the opening paragraph with "ES modules" feels to me like we are going a bit too fast, especially for beginners who've never done anything with JavaScript.
Why would they need to use JavaScript and why would they need to package it?
I was wondering how we introduce Javascript to beginners in the "Getting Started" guides.
The JS section with:
> CSS & JavaScript are core to building web applications, so let's learn how to use them with Rails.

Also, the Asset Pipeline is currently more detailed on describing requiring JavaScript than this guide. For example the [Asset Load Order section](https://edgeguides.rubyonrails.org/asset_pipeline.html#asset-load-order) describes the `javascript_include_tag` which  this guide currently doesn't. This is still a supported method for requiring JavaScript. Explaining the older `javascript_include_tag` is useful for readers working on older Rails applications (still using ) as well. 

We could make
Starting this guide with a simple "Hello, world" script in `app/javascript/main.js` (or `app/javascript/hello.js`). This would allow explaining where to place javascript (to help 

This would allow us to mention why you'd want to use ES modules for more complex JS code.

Using a JavaScript Import Map
-----------------------------

A JavaScript
[import map](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/script/type/importmap)
allows JavaScript files to be delivered separately without bundling and still be
able to reference each other.

It is a JSON object which defines the mapping between the module specifier
passed to an `import` statement, and the path to the actual file to be imported.
Here's an example:

```json
{
  "imports": {
    "animations": "/scripts/animations.js",
    "utilities": "/scripts/utilities.js"
  }
}

> ```

Reviewer: Reading the text it's a bit unclear how this JSON is delivered to the browser.
Maybe include the script tag as well? (This is also what the Mozilla documentation considers part of the importmap it seems).
````suggestion
```html
<script type="importmap" data-turbo-track="reload">{
  "imports": {
    "animations": "/scripts/animations.js",
    "utilities": "/scripts/utilities.js"
  }
}
</script>
```
````

browser will import the corresponding file and the module it contains.

In Rails, the import map is constructed using the
[importmap-rails](https://github.com/rails/importmap-rails) gem.

This technique doesn't require an additional build step for your JavaScript.
Your files are delivered as-is, and hence no JavaScript runtime such as Node.js
is required.

### Installing `importmap-rails`

The `importmap-rails` gem is included by default in all new Rails applications.
In older applications, you can install it using:

```bash
$ bundle add importmap-rails
$ bin/rails importmap:install
```

### Declaring JavaScript Files

All your JavaScript files need to be declared in `config/importmap.rb` so Rails
knows to include them in the import map object.

```ruby
# config/importmap.rb

# Declare JavaScript files from your application
pin "application"
pin "utilities"

# Declare all files inside a folder
pin_all_from "app/javascript/controllers", under: "controllers"
```

NOTE: All files declared in your `config/importmap.rb` must exist within your
[asset pipeline's load paths](asset_pipeline.html#load_paths).

This will create an import map object similar to:

> ```json

Reviewer: Can we show the full <script> output here as well, as that is what `javascript_importmap_tags` outputs?

{
  "imports": {
    "application": "/assets/application-d8a8613a.js",
    "utilities": "/assets/utilities-e8dc057d.js",
    "controllers/application": "/assets/controllers/application-3affb389.js",
    "controllers/hello_controller": "/assets/controllers/hello_controller-708796bd.js",
    "controllers": "/assets/controllers/index-ee64e1f1.js"
  }
}
```

which is rendered in your HTML document's `<head>` using:

```erb
<%= javascript_importmap_tags %>

> ```

Reviewer: Nitpick. This somewhat breaks the guide default of starting examples with the code followed by the generated results.

Author: True, but in this case I think the narrative flows better as the reader can mentally make the connection between the import map config and what it produces. The `javascript_importmap_tags` line is boilerplate and logically unimportant.

NOTE: You'll notice that the filenames contain a _hash_. This is added by
[Rails' Asset Pipeline](asset_pipeline.html). It is calculated based on the
file's contents and used to version the files.

> 

Reviewer: It would be nice if we can show a complete example here of importing a npm package and using it in the Rails application.
What do you think about duplicating the local-time example, as shown on the [Rails homepage video](https://videos.37signals.com/rails/rails-8-demo.mp4) at 11:45?

There's also the [React example](https://www.youtube.com/watch?v=k73LKxim6tw) but I'm not sure that is a typical use case for importmaps-rails.

Author: We cover that in detail in the asset pipeline guide. A lot of this guide treads on the toes of that guide and I'm not keen on adding more duplication to be honest. Let me have a think about this though ...

Author: EDIT: We show how to do this further down, I think it works well there.

Reviewer: I would at least expect to see how the `utilities.js` could be integrated in an app.
For example the example above adds a "utilities.js" without mentioning where that is located or how it can be imported.

I reproduced the example with a file created `app/javascript/utilities.js` (`public/utilities.js` didn't work):
```js
// app/javascript/utilities.js
console.log('hi from utilities')
```

This didn't log to the console until I added an import to `application.js`:
```js
import "@hotwired/turbo-rails"
import "controllers"
import "utilities"
```
I think this section would benefit from just a little more explanation.

See the [Asset Pipeline guide](asset_pipeline.html#javascript-import-map) and
the [`importmap-rails` Readme](https://github.com/rails/importmap-rails) for
further information.

Using a JavaScript Bundler
--------------------------

You can integrate a JavaScript bundler into Rails using the
[`jsbundling-rails`](https://github.com/rails/jsbundling-rails) gem. It supports

> a number of builders such as [ESBuild](https://esbuild.github.io/),

Reviewer: ```suggestion
a number of bundlers such as [ESBuild](https://esbuild.github.io/),
```

[Rollup](https://rollupjs.org/guide/en/), [Bun](https://bun.sh), and
[Webpack](https://webpack.js.org/).

> This gem requires a JavaScript runtime. For all bundlers except Bun, you'll need

Reviewer: To avoid confusion maybe mention `jsbundling-rails` explicitly here (a reader might think the bundlers are gems as well)
```suggestion
`jsbundling-rails` requires a JavaScript runtime. For all bundlers except Bun, you'll need
```

> Node.js and Yarn. For Bun, you'll just need to install that as it is both a

Reviewer: The Yarn version mentioned is pretty old and npm comes installed by default with node.js.
jsbundling-rails detects what bundler is used: https://github.com/rails/jsbundling-rails/blob/6ead108f3670b4fd937f38296fdb0a689c7d5df9/lib/tasks/jsbundling/build.rake#L49-L61

I think we can skip installing Yarn as jsbundling-rails will fallback to npm.
We should probably also mention the package manager is detected to avoid confusion why a certain package manager is chosen? (If bun is installed it will default to that).

Author: Yep I agree, changed the wording

JavaScript runtime and a bundler.

### Installing Node.js and Yarn

Find the installation instructions on the
[Node.js website](https://nodejs.org/en/download/) and verify it’s installed

> correctly:

Reviewer: We use `mise` for installing Ruby, not sure we should use that here as well?

Author: I'm not keen on an official guide demonstrating third party tools that are dependent on a user's preferences. For example, I like to use `rv` for Ruby and `nvm` for Node, and dislike tools that try to manage everything.

Reviewer: What I meant is that we use it in the installing Ruby guide: https://guides.rubyonrails.org/install_ruby_on_rails.html#install-ruby-on-macos
But I see that guide uses `wsl` on Windows.
Describing the installation for all OS-es is probably too detailed for Node.js (and a hassle to maintain), so linking to the Node.js website is probably the right approach.

```bash
$ node --version
v23.6.1
```

To install Yarn, follow the installation instructions at the
[Yarn website](https://classic.yarnpkg.com/en/docs/install). Verify it's
installed using:

```bash
$ yarn --version
1.22.19
```

### Installing Bun

Follow the installation instructions at the [Bun website](https://bun.sh) and
verify it’s installed:

```bash
$ bun --version
v1.3.13
```

### Installing `jsbundling-rails`

When creating a new Rails app, setup a JavaScript bundler using the `-j` or
`--javascript` flag:

```bash
$ rails new my_new_app -j esbuild
```

```bash
$ rails new my_new_app --javascript=esbuild
```

Add the `jsbundling-rails` gem in an existing Rails app using:

```bash
$ bundle add jsbundling-rails
```

Then configure your chosen bundler with:

```bash
$ bin/rails javascript:install:[bun|esbuild|rollup|webpack]
```

When using `jsbundling-rails`, use `bin/dev` to start the JavaScript bundler
along with Rails server in development. Further information is available in the
[Asset Pipeline guide](asset_pipeline.html#bundling-and-transpiling-javascript).

> Adding npm Packages

Reviewer: I'm not sure we should have a separate top section for adding  npm packages only.
All the package managers just have a single command to add a package to package.json, that is well documented elsewhere. `importmap-rails` has it's own commands, but is that enough to warrant a separate section?

Also, `bin/importmaps` has a lot more command like `update`, `pristine` and `unpin`, that aren't mentioned yet.

What do you think about moving this section under the "Using a JavaScript Import Map" section and explaining some other common commands like `unpin`?

Author: I don't think this guide should turn into a "using import maps with Rails" guide. If we're going to show how to install npm packages with import maps (which we should because it's the most basic task), we should show how to do it with a bundler too.

I think it should be a top level section because it's one of the core tasks someone might want to do, and it should show in the index pane as such.

Also, the import-maps readme documents all the other commands, and that's linked in this section so users can dig deeper for more information. 

I don't want to replicate that Readme in this guide too much, as it will be tricky to maintain.

-------------------

> ### Vendoring NPM Packages with `importmap-rails`

Reviewer: It looks a bit weird here but the [npm is always lowercase](https://en.wikipedia.org/wiki/Npm).
See also the [naming guidelines](https://edgeguides.rubyonrails.org/api_documentation_guidelines.html#naming).
```suggestion
### Vendoring npm Packages with `importmap-rails`
```

> When using `importmap-rails`, NPM packages are downloaded into the `vendor`

Reviewer: ```suggestion
When using `importmap-rails`, npm packages are downloaded into the `vendor`
```

folder in your app and checked into source control.

Add a package to your application using `bin/importmap pin`:

```bash
$ bin/importmap pin ahoy.js
```

This will download the package into your `vendor` folder and declare them in
your `config/importmap.rb`. You can then import the package in your
`application.js`:

> ```javascript

Reviewer: ````suggestion
```javascript
// app/javascript/application.js
````

import ahoy from "ahoy.js";
```

Further information is available in the
[`importmap-rails` Readme](https://github.com/rails/importmap-rails?tab=readme-ov-file#using-npm-packages-via-javascript-cdns).

> ### Installing NPM Packages with a JavaScript Bundler

Reviewer: Unlike `bin/importmap` these package manager commands aren't part of Rails.
I don't think we should describe them here. Anyone using a package manager probably knows how to use that package managers CLI and their usage is well documented elsewhere.

Author: You're right, but it's the most basic demonstration. I want to keep it there for completeness even thought they're not part of Rails. It'd be useful to the reader and it doesn't cover any detail, and there's a link to the official docs for further information.

I think it keeps the guide beginner friendly and helps with the flow to have just a couple of paragraphs explaining this, even if it's, strictly speaking, not part of Rails.

When using Bun, the Bun package manager installs NPM packages:

```bash
$ bun add ahoy.js
```

See the [Bun documentation](https://bun.com/docs/pm/cli/install) for more
information.

For all other bundlers, use Yarn to manage your dependencies:

```bash
$ yarn add ahoy.js
```

Further details are available in the
[Yarn documentation](https://yarnpkg.com/getting-started/usage).

> Choosing Between an Import Map and a JavaScript Bundler

Reviewer: This section has a lot of overlap with the [Advanced Asset Management](https://edgeguides.rubyonrails.org/asset_pipeline.html#advanced-asset-management) section in the  "Asset pipeline" guide. That section is a lot more detailed and also mentions CSS assets. So I think it makes sense to keep that, but we have some duplication now.

Should we link there instead?

Author: There is a bit of duplication between this guide and the asset pipeline guide. It's tricky to know what to keep to be honest, and I don't want this guide to turn into just a bunch of external links.

In my mind, if something comes under the most basic questions a new user may have relating to JavaScript in Rails, it belongs in this guide. As such I'd rather keep this section here as is.

Reviewer: Yeah, I think I agree.

-------------------------------------------------------

In all new Rails apps, JavaScript is delivered using an import map. The Rails

> team believes that using an import maps reduces complexity, improves developer

Reviewer: ```suggestion
team believes that using an import map reduces complexity, improves developer
```

Introduction
------------

ES modules (ESM) is the standard structure for packaging JavaScript
code in modern applications. Rails provides two mechanisms to
deliver JavaScript made for ESM: an
[import map](#using-a-javascript-import-map), or a
[JavaScript bundler](#using-a-javascript-bundler). Both of these systems
integrate with the [Asset Pipeline](asset_pipeline.html) to deliver
the files to the browser.

The [Turbo](#turbo) and [Stimulus](#stimulus) JavaScript libraries (which are
part of the [Hotwire](https://hotwired.dev) suite) are included in Rails, and
form the default front-end stack.

Using a JavaScript Import Map
-----------------------------

A JavaScript
[import map](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/script/type/importmap)
allows JavaScript files to be delivered separately without bundling and still be
able to reference each other.

It is a JSON object which defines the mapping between the module specifier
passed to an `import` statement, and the path to the actual file to be imported.
Here's an example:

```html
<script type="importmap" data-turbo-track="reload">
{
  "imports": {

>     "animations": "/scripts/animations.js",
>     "utilities": "/scripts/utilities.js"

Reviewer: Can we change the directories here?
A Rails developer might confuse these with Rails' `/script` directory.
```suggestion
    "animations": "/javascript/animations.js",
    "utilities": "/javascript/utilities.js"
```

When creating a new Rails app, setup a JavaScript bundler using the `-j` or
`--javascript` flag:

```bash
$ rails new my_new_app -j esbuild
```

```bash
$ rails new my_new_app --javascript=esbuild
```

Add the `jsbundling-rails` gem in an existing Rails app using:

```bash
$ bundle add jsbundling-rails
```

Then configure your chosen bundler with:

```bash
$ bin/rails javascript:install:[bun|esbuild|rollup|webpack]
```

When using `jsbundling-rails`, use `bin/dev` to start the JavaScript bundler
along with Rails server in development. Further information is available in the
[Asset Pipeline guide](asset_pipeline.html#bundling-and-transpiling-javascript).

Adding npm Packages
-------------------

### Vendoring npm Packages with `importmap-rails`

When using `importmap-rails`, npm packages are downloaded into the `vendor`
folder in your app and checked into source control.

Add a package to your application using `bin/importmap pin`:

```bash
$ bin/importmap pin ahoy.js
```

This will download the package into your `vendor` folder and declare them in
your `config/importmap.rb`. You can then import the package in your
`application.js`:

```javascript
// app/javascript/application.js

> import ahoy from "ahoy.js";

Reviewer: This example was a bit confusing to me as I would expect the application would use `ahoy` to make some javascript calls.

Even just some configuration:
```
ahoy.configure({page: "Landing page"});
```

Or maybe just some explanation above the example that explains what `ahoy.js` is used for and what the import executes in this case.

action reads the `name` _target_ and writes it into the `output` _target_.

It can be attached to the DOM via the `data-controller` attribute:

```html
<div data-controller="hello">
  <input data-hello-target="name" type="text">

  <button data-action="click->hello#greet">Greet</button>

  <span data-hello-target="output"> </span>
</div>
```

Refer to the Stimulus
[handbook](https://stimulus.hotwired.dev/handbook/introduction) and
[reference](https://stimulus.hotwired.dev/reference/controllers) for complete
usage information.

Rails integrates with Stimulus via the [`stimulus-rails`][] gem, which provides
a generator to create Stimulus controllers:

```bash
# Generates app/javascript/controllers/hello_controller.js
$ bin/rails generate stimulus hello
```

```js
// app/javascript/controllers/hello_controller.js

import { Controller } from "@hotwired/stimulus"

// Connects to data-controller="hello"
export default class extends Controller {
  connect() {}
}
```

It also contains a _task_ which you can use to install Stimulus in an existing
application:

```bash
$ bin/rails stimulus:install
```

[`stimulus-rails`]: https://github.com/hotwired/stimulus-rails

> ### `request.js`

Reviewer: Nitpick. request.js can be used outside Hotwire.
We should keep this section if we remove the Hotwire section from this guide.

Author: Yep I've kept this section here for that very reason :)
